### PR TITLE
Element wise optimizations

### PR DESCRIPTION
**Pull request by vivek-rane**
Added MKL element-wise ops that utilize eigen ops as their back-end. Also added an input-conversion op that ensures that shapes of both input tensors are compatible (same or broadcastable).

### CODE REVIEW

**Comment by tfboyd** [comment]
Ping @andydavis1 @rmlarsen

File: tensorflow/core/graph/mkl_layout_pass.cc [hunk 1]
```diff
@@ -1956,7 +1984,31 @@ MklLayoutRewritePass::CheckForNodeRewrite(const Node* n) const {
 
   // BiasAddGrad is not an Mkl layer, so we make an exception for it.
   if (n->type_string() != csinfo_.bias_add_grad) {
-    if (!mkl_op_registry::IsMklOp(GetMklOpName(n->type_string()), T)) {
+    if (!mkl_op_registry::IsMklOp(mkl_op_registry::GetMklOpName(n->type_string()), T)) {
+      return nullptr;
+    }
+  }
+
+  // For elementwise node, we reuse the Eigen implementation and pass the MKL
```
Comment (andydavis1) [inline tensorflow/core/graph/mkl_layout_pass.cc]: Please add a TODO to add your own implementation in the future so we can avoid inter-library dependencies.

**Review by andydavis1 — APPROVED** [review]

File: tensorflow/core/graph/mkl_layout_pass.cc (hunk 1 above)
Comment (vivek-rane) [reply tensorflow/core/graph/mkl_layout_pass.cc]: Done.

**Comment by vivek-rane** [comment]
@tfboyd @andydavis1 @rmlarsen Please merge this one second (after 12654)

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
MacOS rerun: http://ci.tensorflow.org/job/tensorflow-pull-requests-mac/6176

**Comment by vivek-rane** [comment]
From the rerun:

```
Executed 1193 out of 1193 tests: 1193 tests pass.

Parameterized build ends with SUCCESS at: Tue Sep  5 09:51:01 PDT 2017 (Elapsed time: 2281 s)
Finished: SUCCESS
```

**Comment by vivek-rane** [comment]
@tfboyd @andydavis1 @rmlarsen @martinwicke any reason we're blocked on this one?

**Comment by tfboyd** [comment]
Let me get double check.  It looks like Martin had the Mac tests rerun and they are fine.

**Comment by tfboyd** [comment]
@yifeif   Can you take a look and get this merged if everything is ok?  Thank you very much.  We are ok with the changes as long as the tests pass.

**Comment by vivek-rane** [comment]
@yifeif sorry to pester but we have a lot of things waiting on this pull request. Please can you check if merge is possible? (see comment by @tfboyd)

**Comment by yifeif** [comment]
We are ready to merge now! Thanks for the patience.

**Pull request merged**